### ISSUE TITLE
windows下无法正常编译安装

### ISSUE THREAD

**Issue by SimpleRealDT** [OpenBMB/BMTrain#54]
pip install bmtrain 
Collecting bmtrain
  Using cached bmtrain-0.1.8.post1.tar.gz (48 kB)
  Preparing metadata (setup.py) ... done
Requirement already satisfied: numpy in c:\users\chenliyu\anaconda3\envs\cpm\lib\site-packages (from bmtrain) (1.21.6)
Building wheels for collected packages: bmtrain
  Building wheel for bmtrain (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [55 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-37
      creating build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\backward.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\block_layer.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\checkpointing.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\debug.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\global_var.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\init.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\layer.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\parameter.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\param_init.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\store.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\synchronize.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\utils.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\wrapper.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain
      creating build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\all_gather.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark    
      copying bmtrain\benchmark\reduce_scatter.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\shape.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\utils.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark      
      creating build\lib.win-amd64-cpython-37\bmtrain\distributed
      copying bmtrain\distributed\ops.py -> build\lib.win-amd64-cpython-37\bmtrain\distributed       
      copying bmtrain\distributed\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\distributed  
      creating build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\format.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\model.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\tensor.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      creating build\lib.win-amd64-cpython-37\bmtrain\loss
      copying bmtrain\loss\cross_entropy.py -> build\lib.win-amd64-cpython-37\bmtrain\loss
      copying bmtrain\loss\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\loss
      creating build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\cosine.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\exponential.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\linear.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\noam.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\no_decay.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\warmup.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      creating build\lib.win-amd64-cpython-37\bmtrain\nccl
      copying bmtrain\nccl\enums.py -> build\lib.win-amd64-cpython-37\bmtrain\nccl
      copying bmtrain\nccl\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\nccl
      creating build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\adam.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\adam_offload.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\clip_grad.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      running build_ext
      error: [WinError 2] 系统找不到指定的文件。
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for bmtrain
  Running setup.py clean for bmtrain
Failed to build bmtrain
Installing collected packages: bmtrain
  Running setup.py install for bmtrain ... error
  error: subprocess-exited-with-error

  × Running setup.py install for bmtrain did not run successfully.
  │ exit code: 1
  ╰─> [57 lines of output]
      running install
      C:\Users\chenliyu\anaconda3\envs\cpm\lib\site-packages\setuptools\command\install.py:37: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        setuptools.SetuptoolsDeprecationWarning,
      running build
      running build_py
      creating build
      creating build\lib.win-amd64-cpython-37
      creating build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\backward.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\block_layer.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\checkpointing.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\debug.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\global_var.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\init.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\layer.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\parameter.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\param_init.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\store.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\synchronize.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\utils.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\wrapper.py -> build\lib.win-amd64-cpython-37\bmtrain
      copying bmtrain\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain
      creating build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\all_gather.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\reduce_scatter.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\shape.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\utils.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      copying bmtrain\benchmark\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\benchmark
      creating build\lib.win-amd64-cpython-37\bmtrain\distributed
      copying bmtrain\distributed\ops.py -> build\lib.win-amd64-cpython-37\bmtrain\distributed
      copying bmtrain\distributed\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\distributed
      creating build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\format.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\model.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\tensor.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      copying bmtrain\inspect\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\inspect
      creating build\lib.win-amd64-cpython-37\bmtrain\loss
      copying bmtrain\loss\cross_entropy.py -> build\lib.win-amd64-cpython-37\bmtrain\loss
      copying bmtrain\loss\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\loss
      creating build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\cosine.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\exponential.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\linear.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\noam.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\no_decay.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\warmup.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      copying bmtrain\lr_scheduler\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\lr_scheduler
      creating build\lib.win-amd64-cpython-37\bmtrain\nccl
      copying bmtrain\nccl\enums.py -> build\lib.win-amd64-cpython-37\bmtrain\nccl
      copying bmtrain\nccl\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\nccl
      creating build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\adam.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\adam_offload.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\clip_grad.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      copying bmtrain\optim\__init__.py -> build\lib.win-amd64-cpython-37\bmtrain\optim
      running build_ext
      error: [WinError 2] 系统找不到指定的文件。

已经将cl加入环境变量，python版本为3.7，已更新vc++build tool

**Comment by SimpleRealDT** [comment]
done

**Comment by Beaujor** [comment]
I meet same problem. can you tell me why?

**Comment by SimpleRealDT** [comment]
> I meet same problem. can you tell me why?

It works fine under ubuntu, you need to make sure you have the full nvidia cuda installed not anaconda codatoolkit, and configure the correct environment variables, the compilation fails because the cuda source cannot be found
